### PR TITLE
Handle interval queries against hg38 contigs. Detect and reject ambiguous query intervals.

### PR DESCRIPTION
**Pull request by cmnbroad**
Fixes https://github.com/broadinstitute/gatk/issues/1438. It is still in theory possible to have queries that are ambiguous that will be rejected (though not with hg38). Longer term we'll need to implement a mechanism for quoting or otherwise allowing the user to handle how to resolve those.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java [hunk 1]
```diff
@@ -4,10 +4,16 @@
 import htsjdk.samtools.SAMSequenceDictionary;
 import htsjdk.samtools.SAMSequenceRecord;
 import htsjdk.samtools.util.Locatable;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.broadinstitute.hellbender.exceptions.GATKException;
 import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.read.mergealignment.SamAlignmentMerger;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: Remove stray import

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java [hunk 2]
```diff
@@ -135,12 +142,124 @@ public SimpleInterval(final String str){
...
+      * {@code sequenceDictionary}. If the list is empty, the query doesn't match any contig in the sequence
+      * dictionary. If the list contains more than one interval, the query string is ambiguous and should be
+      * rejected. If the list contains a single interval, the query is unambiguous and can be safely used to
+      * conduct a query.
+      * @thows IllegalArgumentException if the query only matches a single contig in the dictionary, but the start
+      * and end positions are not valis
+      * @throws NumberFormatException if the query only matches a single contig in the dictionary, but the query
+      * interval paramaters (start, end) cannot be parsed
+      */
+     public static List<SimpleInterval> getResolvedIntervals(
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: Can this be moved to `IntervalUtils`? Have this huge utility method directly in `SimpleInterval` like this is a little gross...  (if it would be non-trivial we can move it in a separate PR).

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java [hunk 3]
```diff
@@ -135,12 +142,124 @@ public SimpleInterval(final String str){
...
+             final String intervalQueryString,
+             final SAMSequenceDictionary sequenceDictionary) {
+         Utils.nonNull(intervalQueryString);
+         Utils.validateArg(!intervalQueryString.isEmpty(), "intervalQueryString should not be empty");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: Sequence dictionary must also be non-null

File: src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java [hunk 4]
```diff
@@ -504,4 +505,21 @@ public void testGenomeLocPadding(final GenomeLoc input, final int pad) {
         Assert.assertEquals(padded.getStop(), Math.min(input.getStop() + pad, contigLength));
     }
 
+    @Test
+    public void testQueryAllHG38Intervals() {
+        SAMSequenceDictionary sd;
+        final File testFile = new File ("src/test/resources/org/broadinstitute/hellbender/engine/Homo_sapiens_assembly38.headerOnly.vcf.gz");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java]: Use inherited `publicTestDir` constant

File: src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java [hunk 5]
```diff
@@ -504,4 +505,21 @@ public void testGenomeLocPadding(final GenomeLoc input, final int pad) {
...
+        try (VCFFileReader vcfReader = new VCFFileReader(testFile, false)) {
+            sd = vcfReader.getFileHeader().getSequenceDictionary();
+        }
+
+        // Test that we can use any contig from hg38 as a query against a VCF with an hg38 sequence dictionary without
+        // having an ambiguity.
+        final GenomeLocParser localGenomeLocParser = new GenomeLocParser(sd);
+        sd.getSequences().stream().forEach(hg38Contig -> localGenomeLocParser.parseGenomeLoc(hg38Contig.getSequenceName()));
+        sd.getSequences().stream().forEach(hg38Contig -> localGenomeLocParser.parseGenomeLoc(hg38Contig.getSequenceName() + ":1-1"));
+        sd.getSequences().stream().forEach(hg38Contig -> localGenomeLocParser.parseGenomeLoc(hg38Contig.getSequenceName() + ":1+"));
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java]: Add a test for the "contig:1" case here. Also, assert that the return value of `parseGenomeLoc()` is as expected

File: src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java [hunk 6]
```diff
@@ -439,6 +440,149 @@ public void testParseIntervalArguments() {
...
+                    ambiguousIntervals = SimpleInterval.getResolvedIntervals(
+                            new SimpleInterval(hg38Contig.getSequenceName(), 1, hg38Contig.getSequenceLength()).toString(), sd);
+                    Assert.assertNotNull(ambiguousIntervals);
+                    Assert.assertEquals(ambiguousIntervals.size(), 1);
+
+                    // query using an interval with the format "name:1+"
+                    ambiguousIntervals = SimpleInterval.getResolvedIntervals(hg38Contig.getSequenceName() + ":1+", sd);
+                    Assert.assertNotNull(ambiguousIntervals);
+                    Assert.assertEquals(ambiguousIntervals.size(), 1);
+                }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java]: Also test the ":1" case, and check the contents of `ambiguousIntervals` to make sure the interval it contains is correct.

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java (hunk 1 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: Done.

File: src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java (hunk 5 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java]: Done.

File: src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java (hunk 6 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/utils/IntervalUtilsUnitTest.java]: Done.

File: src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java (hunk 4 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/utils/GenomeLocParserUnitTest.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java (hunk 2 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java]: Done.

**Comment by cmnbroad** [comment]
Also rebased.

**Comment by jonn-smith** [comment]
Nothing jumps out at me (modulo the changes droazen suggested).  I think it should be fine.

**Comment by droazen** [comment]
Go ahead and merge @cmnbroad :+1:
